### PR TITLE
Reduce PML memory consumption

### PR DESCRIPTION
**Pull request by sbastrakov**
By internally storing only the required near-boundary values instead of the whole local domain as for normal E and B. Along the way did small style improvements.

TODO:

- [x] fix 2D case
- [x] test checkpointing
- [x] refactor and document
- [x] update the memory calculator

### CODE REVIEW

File: .build/CMakeFiles/3.12.3/CMakeSystem.cmake [hunk 1]
```diff
@@ -0,0 +1,15 @@
+set(CMAKE_HOST_SYSTEM "Windows-10.0.14393")
```
Comment (ax3l) [inline .build/CMakeFiles/3.12.3/CMakeSystem.cmake]: feel free to add the `.build` directory to `.gitignore` if that's a thing that VS usually has in place.

Comment (sbastrakov) [reply .build/CMakeFiles/3.12.3/CMakeSystem.cmake]: Oops, thx for pointing out.

Comment (sbastrakov) [reply .build/CMakeFiles/3.12.3/CMakeSystem.cmake]: That was actually not VS, just me mimicking the usual directory structure.

**Comment by sbastrakov** [comment]
Ready for review.
Note: the workarounds for the HDF5 and ADIOS output plugins are mostly similar, could have been abstracted better, and implemented with less duplication. After offline discussions with @psychocoderHPC we decided not to do it, and just implement patches so that it all works for now. The reason is that hopefully these 2 plugins will be rather soon replaced with a new openPMD plugin developed in #2966 (which perhaps needs to be extended to support non-domain-bound fileds, to be discussed there).

**Comment by sbastrakov** [comment]
Sorry for accidentally introducing other commits, I did rebase inaccurately, now should be fixed.
